### ISSUE TITLE
HttpClient super slow when used with localhost

### ISSUE THREAD

**Issue by brhoten** [dotnet/corefx#39093]
I'm using dotnet-sdk-3.0.100-preview6-012264-win-x64, but experienced the same problem in older .net core versions. Does not appear to be a problem with .net framework implementations.

```var content = await client.PostAsync("http://localhost:8187/", payload);```

is about 10-100x slower than

```var content = await client.PostAsync("http://127.0.0.1:8187/", payload);```

**Comment by stephentoub** [comment]
@brhoten, just for the first request, or for all requests?

What OS distro / version are you using?

What is the payload?  Can you share a complete repro?

**Comment by brhoten** [comment]
payload was a JSON-ified string. The slow down was evident on every iteration over the PostAsync call.

Edition: Windows 10 Pro
Version: 1803
OS Build: 17134.829

A cleaned up version of the code looks like this, but The only difference in the two attempts was the above change from `localhost` to `127.0.0.1` in the PostAsync. 

```var client = new HttpClient();

	foreach (var datum in data)
	{
	    var payload = new StringContent( JsonSerializer.ToString(datum.Raw));
	    var content = await client.PostAsync("http://localhost:8187/", payload);
	    var result = await JsonSerializer.ReadAsync<CustomClass[]>(await content.Content.ReadAsStreamAsync());
	    ...
	}```

**Comment by stephentoub** [comment]
Do you have a proxy configured that might be trying to resolve "localhost"?

Can you capture a network trace that might show what's going on?

I just tried putting together a little repro, and couldn't reproduce what you're seeing.  I run this:
```C#
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

public class CustomClass
{
    public int Value1 { get; set; }
    public string Value2 { get; set; }
}

class Program
{
    static async Task Main()
    {
        var data = new CustomClass[] { new CustomClass { Value1 = 42, Value2 = "84" }, new CustomClass { Value1 = 1, Value2 = "2" } };
        string payload = JsonSerializer.Serialize(data) + "\r\n";
        byte[] responseBytes = Encoding.ASCII.GetBytes("HTTP/1.1 200 OK\r\nContent-Length: " + payload.Length + "\r\n\r\n" + payload);

        using (var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
        {
            listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
            listener.Listen(1);
            var uriLocalhost = new Uri("http://localhost:" + ((IPEndPoint)listener.LocalEndPoint).Port);
            var uri127001 = new Uri("http://127.0.0.1:" + ((IPEndPoint)listener.LocalEndPoint).Port);

            _ = Task.Run(() =>
            {
                while (true)
                {
                    Socket server = listener.Accept();
                    Task.Run(() =>
                    {
                        using (server)
                        using (var reader = new StreamReader(new NetworkStream(server, true)))
                        {
                            while (true)
                            {
                                while (!string.IsNullOrEmpty(reader.ReadLine())) ;
                                reader.ReadLine();
                                server.Send(responseBytes);
                            }
                        }
                    });
                }
            });

            using (var client = new HttpClient())
            {
                async Task RunAsync(int iterations, Uri uri)
                {
                    for (int i = 0; i < iterations; i++)
                    {
                        using (HttpResponseMessage r = await client.PostAsync(uri, new StringContent(payload)))
                        using (Stream s = await r.Content.ReadAsStreamAsync())
                        {
                            await JsonSerializer.DeserializeAsync<CustomClass[]>(s);
                        }
                    }
                }

                // Warmup
                await RunAsync(1, uriLocalhost);
                await RunAsync(1, uri127001);

                // Test
                var sw = new Stopwatch();
                while (true)
                {
                    sw.Restart();
                    await RunAsync(10_000, uri127001);
                    Console.WriteLine("127.0.0.1: " + sw.Elapsed);

                    sw.Restart();
                    await RunAsync(10_000, uriLocalhost);
                    Console.WriteLine("localhost: " + sw.Elapsed);

                    Console.WriteLine();
                }
            }
        }
    }
}
```
and get results like:
```
127.0.0.1: 00:00:00.6861562
localhost: 00:00:00.6679440

127.0.0.1: 00:00:00.6062265
localhost: 00:00:00.4863958

127.0.0.1: 00:00:00.4869082
localhost: 00:00:00.4834639

127.0.0.1: 00:00:00.4865064
localhost: 00:00:00.4845890

127.0.0.1: 00:00:00.4905811
localhost: 00:00:00.4827975

127.0.0.1: 00:00:00.4859724
localhost: 00:00:00.4771593

127.0.0.1: 00:00:00.4963004
localhost: 00:00:00.4929136
...
```

You see otherwise?

**Comment by brhoten** [comment]
Strange.

I had to change your serialization call to `ToString` and the deserialization to `ReadAsync', but I'm getting similar performance in this set up.

```127.0.0.1: 00:00:00.0112476
localhost: 00:00:00.0104458

127.0.0.1: 00:00:00.0098451
localhost: 00:00:00.0105029

127.0.0.1: 00:00:00.0095863
localhost: 00:00:00.0098702

127.0.0.1: 00:00:00.0093019
localhost: 00:00:00.0085527

127.0.0.1: 00:00:00.0092219
localhost: 00:00:00.0080159

127.0.0.1: 00:00:00.0118243
localhost: 00:00:00.0130008

127.0.0.1: 00:00:00.0168641
localhost: 00:00:00.0210877

127.0.0.1: 00:00:00.0167247
localhost: 00:00:00.0135452```

I spent most of the afternoon on this and was seeing similar performance to what I earlier reported until I recopied your test above and tried it minimally modified. I disabled the virtual docker network that was running in the background, disabled firewall, and shut off WSL hosting Ubuntu (this is where the original service is hosted). None seemed to make a difference. Not sure why it's working all of a sudden.

I still get the same low performance in other code, but yours seems to work fine.

**Comment by davidsh** [comment]
@brhoten Thanks for confirming that the problem is probably due to your network environment and not the .NET Core software. 

Sometimes these problems are due to slow DNS resolution by outside providers (i.e. your Internet ISP) which might try to resolve "localhost". 

It would be best if you could isolate a repro for this and attach a network trace to show the timings of where things are slowing down. WireShark is a good tool for this.

In the meantime, we will close this issue. Please let us know if you are able to get more detailed information that might point to a possible bug in .NET Core.

**Comment by brhoten** [comment]
I don't know that I can let it go that easily. Under .Net Framework 4.8 I can alternate `localhost` and `127.0.0.1` without any problem. It is only under .Net Core that I see this behavior. That would still point to something different about the implementation under .Net Core. 

I ran a couple of wireshark captures between the test code above talking to the in-process server it creates, the above test code talking to my server, my original client with both `localhost` and `127.0.0.1` talking to my server, and finally my original client code on .Net Framework 4.8 talking to my server.

The server is an WSL Ubuntu instance hosted on the same machine. I noticed that the .Net Core with `localhost` seems to get a lot of errors while trying to invoke the IPv6 loopback address `::1`. The others do not have this same problem.

I'm willing to say this isn't a normal setup, but when the only differing factor is `127.0.0.1` and `localhost` between .Net Core and .Net Framework and the localhost causes an exponentially longer lag time .Net Core it should still point to something in the library.
